Fix client potentially not leaving a room on exiting multiplayer

smoogipoo opened the pull request:
Alternative to / closes #11542
Closes #11315 

I took a different approach by making `JoinRoom()` / `LeaveRoom()` synchronous in regards to each other. This fixes the issue in #11542 which is caused by timing as `LeaveRoom` schedules to the update thread.

I've also removed the `getRoomUsers()` method because it's now done on the `MultiplayerRoom` directly on the async context.

I've tested that it fixes the scenario mentioned in #11542.

smoogipoo committed:
Tie JoinRoom() and PartRoom() together

smoogipoo committed:
Fix potential test failures due to timing

smoogipoo committed:
Remove unnecessary condition blocking the part

smoogipoo committed:
Reset task post-execution

smoogipoo commented on osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs:
```diff
@@ -534,5 +568,25 @@ private void updateUserPlayingState(int userId, MultiplayerUserState state)
                 CurrentMatchPlayingUserIds.Remove(userId);
         }
+
+        private Task scheduleAsync(Action action)
```
I didn't implement this as an extension method because this is quite unsafe due to `async () =>` which probably requires something like https://github.com/ppy/osu-framework/pull/4149 to be handled properly.

peppy replied:
can you explain where the `async () =>` is? having trouble understanding this comment.

peppy replied:
I'm thinking this would be a pretty handy method to have in `Scheduler` itself (or just make our existing scheduler delegate stuff into `Task`s finally)

smoogipoo replied:
There is no `async () =>` here, because all usages have made sure not to do so. My comment was intended to read that _it's possible_ for a user to write such a lambda if this were moved to an extension method, at which point it needs proper handling.

peppy replied:
I see. Would you be against this being added to `Scheduler` itself in some way?

smoogipoo replied:
No I'm not, so long as it's handled properly. I agree it's a useful thing and I originally had it as an extension method before realising its shortcomings.

bdach commented on osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs:
```diff
@@ -109,30 +108,66 @@ protected StatefulMultiplayerClient()
 
-            apiRoom = room;
-            playlistItemId = room.Playlist.SingleOrDefault()?.ID ?? 0;
+                    // Should be thread-safe since joinOrLeaveTask is locked on in both JoinRoom() and LeaveRoom().
+                    if (Room != null)
+                        throw new InvalidOperationException("Cannot join a multiplayer room while already in one.");
```
I'm going to be honest, I'm nowhere near being even 10% sure this is thread-safe given the current complexity of this code.

All current mutations to `Room` are done inside of a schedule inside of a TPL task, and the task to which the field is assigned to is protected by a lock. Which by itself is a mouthful and a half, but the TPL part and schedule part is especially why I'm skeptical this is safe.

As far as I can see the only part that the lock protects is the assignment to `joinOrLeaveTask` - IOW, it makes sure that a leave task can't be "dropped" by a new join and not awaited on, due to it being written over, and vice-versa.

It does *not*, as far as I can see, guarantee that `lastTask` is not running concurrently alongside `newTask`, because the *body* of `Task.Run()` is not in the critical section (and even if it was, it'd still be super ultra bad, because awaits inside critical sections sound like immediate deadlock trouble, and should really fail to compile if I'm not mistaken). In theory:

```
old leave task			new join task

						reads non-null room

executes scheduled portion
clears out room

						continues past check and explodes
```

Am I missing something?

smoogipoo replied:
Am I wrong in thinking that it's the `await lastTask` above which guarantees that it's not running alongside `newTask`?

bdach replied:
I don't think so, as I believe `Task.Run()` will kick off the task potentially immediately? I'm thinking the await on the old task won't help if you don't delay starting the new task after it, but I haven't experimentally confirmed this.

bdach replied:
Never mind, I forgot that the await is inside the tasks. It still feels a little dicey though.

bdach commented on osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs:
```diff
@@ -109,30 +108,66 @@ protected StatefulMultiplayerClient()
+            finally
+            {
+                // The task will be awaited in the future, so reset it so that the user doesn't get into a permanently faulted state if anything fails.
+                lock (joinOrLeaveTask)
```
Should take
```suggestion
                lock (joinOrLeaveTaskLock)
```
right?

bdach commented on osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs:
```diff
@@ -142,21 +177,47 @@ public async Task JoinRoom(Room room)
+            finally
+            {
+                // The task will be awaited in the future, so reset it so that the user doesn't get into a permanently faulted state if anything fails.
+                lock (joinOrLeaveTask)
```
see above

smoogipoo committed:
Fix locking on incorrect object

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

bdach commented on osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs:
```diff
@@ -109,30 +108,66 @@ protected StatefulMultiplayerClient()
+                lastTask = joinOrLeaveTask;
+                joinOrLeaveTask = newTask = Task.Run(async () =>
+                {
+                    if (lastTask != null)
```
Now that I think about it, there might be a theoretical data race here? What I'm thinking is that this `lastTask` read is not under locking, so it might end up reading `newTask`, or even another task entirely, although I'm not sure as I don't quite know how lambda capturing will work in this case...

smoogipoo replied:
It can't, it's a local variable.

peppy commented:
Is there a better way to handle these scenarios without all this complexity overhead? It's very similar to the problem I aimed to fix in https://github.com/ppy/osu/pull/11565.

These are cases where we don't care about threading or concurrency; we just want operations to be run sequentially. I'd really like to hope we are doing things wrong if we need to add these kinds of hard-to-native logical flows each time :/.

bdach commented:
Yeah, my feelings are similar. If you asked me if this code was correct I couldn't conclusively state if it is or isn't, which is probably clear by now seeing I had two swings at finding potential issues and missed both.

The schedule from before was sorta trying to achieve the same goal of serialising the operations, I think? It's unclear to me from the PR descriptions what the actual root cause is, so I don't have anything to suggest without looking into this myself which seems unproductive. I did see a mention of the multiplayer screen stack - is it that operations from within the multiplayer screens get scheduled to the client's scheduler, which continues to run after the multiplayer screen stack is exited, making it possible for a join to happen out of multiplayer?

smoogipoo committed:
Use a task chain and fix potential misordering of events

smoogipoo committed:
Prevent some data races

smoogipoo committed:
Add xmldoc

smoogipoo commented:
I've reworked the task chaining to hopefully be more readable and sane. Please check again.

bdach commented on osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs:
```diff
@@ -534,5 +523,25 @@ private void updateUserPlayingState(int userId, MultiplayerUserState state)
+        {
+            var tcs = new TaskCompletionSource<bool>();
+
+            Scheduler.Add(() =>
+            {
+                try
+                {
+                    action();
+                    tcs.SetResult(true);
+                }
+                catch (Exception ex)
+                {
+                    tcs.SetException(ex);
+                }
+            });
```
Is there a reason to not permit this to potentially run inline like all the other operations around can?

bdach commented:
Overall this does read cleaner but I'm still not sure this is correct.

smoogipoo commented on osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs:
```diff
@@ -142,23 +149,32 @@ public async Task JoinRoom(Room room)
-            {
-                if (Room == null)
-                    return;
+            if (Room == null)
+                return Task.FromCanceled(new CancellationToken(true));
```
Yeah that's a good catch. Ideally everything should be inside the task chain as you said, just have to consider the case of immediate reset as mentioned in the case below.

As far as races on accesses to `Room`, the entire rest of the task chain must fully complete before another task is run from it, including child tasks. It's therefore safe to await on `Room` to be set to null from the update thread within the task chain, as a future `JoinRoom` task in the chain can't run until that completes. I hope that makes sense.

Cancellation tokens are probably the way to go here.

bdach replied:
Oh right I keep forgetting that `scheduleAsync` will block... Yeah the writes should also be ok.

smoogipoo commented:
What sort of breaking are you seeing?/How are you replicating? My tests are:
1. Joining -> leaving -> joining, where the leave operation has a 15s delay.
2. Joining -> leaving -> create, where the leave operation has a 15s delay.
(On two clients)

bdach commented:
I was doing something like:

* Create room
* Exit multiplayer quickly
* Re-enter multiplayer

The created room with one user inside will be there most of the time if the leave happened quickly enough. Again though, this might well be a separate issue with how room creation is written.

peppy reviewed (dismissed):
Overall I'm quite okay with this logic. I'm fine with the direction, so once we confirm it's not breaking in some way (waiting on @bdach's follow-up testing) it should be good to merge.

smoogipoo committed:
Fix TaskChain performing the action in-line, add test

smoogipoo committed:
Make threading even more thread safe

smoogipoo committed:
wip

frenzibyte commented on osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs:
```diff
@@ -142,23 +149,32 @@ public async Task JoinRoom(Room room)
-            {
-                if (Room == null)
-                    return;
+            if (Room == null)
+                return Task.FromCanceled(new CancellationToken(true));
```
I'm not quite experienced with this so this may be a silly comment, but the fact that this null room check and the reset room schedule below are outside of the task chain means there's still a potential for the issue to occur, which is if `LeaveRoom` is invoked while in the middle of executing `JoinRoom`, before `Room` is set to non-null value.

I've [injected both `JoinRoom` and `LeaveRoom` with logs](https://gist.github.com/frenzibyte/d9f45c00736ad10afda3fca307771dc8) and that's what I got when I very quickly left a room after joining it, and the output matches what I was saying above:
```
2021-01-26 16:51:27 [verbose]: [JoinRoom] Beginning join process
2021-01-26 16:51:27 [verbose]: [LeaveRoom] Beginning leaving process
2021-01-26 16:51:27 [verbose]: [LeaveRoom] Room is null, returning...
2021-01-26 16:51:27 [verbose]: [JoinRoom] Joined room server-side
2021-01-26 16:51:27 [verbose]: [JoinRoom] Users populated
2021-01-26 16:51:27 [verbose]: [JoinRoom] New room assigned
2021-01-26 16:51:27 [verbose]: [JoinRoom] Local room settings updated
```

Notice `LeaveRoom` early-returned due to `Room == null`, as it isn't set on `JoinRoom` yet.

The obvious fix would be to move all of this inside the chain, but I still don't get what the attached comment below is referring to exactly, is it about the user will transition to an invalid room screen when pressing "create button" because of `Room != null`? because from the logs I attached above, `JoinRoom` will still set `Room` to a value, causing the create button issue to still persist.

bdach replied:
[you've rediscovered the issue i already pointed out](https://github.com/ppy/osu/pull/11551#discussion_r563969231). there don't seem to be any new changes since that would resolve it.

frenzibyte replied:
Yeah essentially it is the same issue, but I thought that conversation was considered not possible to occur, so I posted a new one writing about it in detail of how it could happen, wasn't sure which is best but at least issue is pointed out.

frenzibyte reviewed (changes requested):
I don't see any other potential issues with this, but don't take my words on that

bdach commented:
> After some more rigorous testing, I believe the TaskChain implementation is still broken.

Do you have any idea/suspicion what might be broken, or some sort of failure scenario that wasn't already mentioned anywhere to go with that comment? Probably better to share if you've got anything to go on, to make sure that effort isn't duplicated.

smoogipoo commented:
I didn't put in an explanation because there's so much...

1. The code (at least first iteration) runs synchronously because of the straight `taskFunc()`.
2. I've had the outer task complete while the inner task is still running with something like `Add(async () => await Task.Delay(15000));`
3. I needed to add cancellation support, discovered the potential reordering of continuations as I mentioned in discord. tl;dr the continuation shouldn't be cancelled but the inner task should be.

And it kind-of just bubbled out of control after that. I've written tests for some simple cases that I used to debug, which may come useful:
```csharp
[TestFixture]
public class TaskChainTest
{
    private TaskChain taskChain;
    private int currentTask;
    private CancellationTokenSource globalCancellationToken;

    [SetUp]
    public void Setup()
    {
        globalCancellationToken = new CancellationTokenSource();
        taskChain = new TaskChain();
        currentTask = 0;
    }

    [TearDown]
    public void TearDown()
    {
        globalCancellationToken?.Cancel();
    }

    [Test]
    public async Task TestChainedTasksRunSequentially()
    {
        var task1 = addTask();
        var task2 = addTask();
        var task3 = addTask();

        task3.mutex.Set();
        task2.mutex.Set();
        task1.mutex.Set();

        await Task.WhenAll(task1.task, task2.task, task3.task);

        Assert.That(task1.task.Result, Is.EqualTo(1));
        Assert.That(task2.task.Result, Is.EqualTo(2));
        Assert.That(task3.task.Result, Is.EqualTo(3));
    }

    [Test]
    public async Task TestChainedTaskWithIntermediateCancelRunsInSequence()
    {
        var task1 = addTask();
        var task2 = addTask();
        var task3 = addTask();

        // Cancel task2, allow task3 to complete.
        task2.cancellation.Cancel();
        task2.mutex.Set();
        task3.mutex.Set();

        // Allow task3 to potentially complete.
        Thread.Sleep(1000);

        // Allow task1 to complete.
        task1.mutex.Set();

        // Wait on both tasks.
        await Task.WhenAll(task1.task, task3.task);

        Assert.That(task1.task.Result, Is.EqualTo(1));
        Assert.That(task2.task.IsCompleted, Is.False);
        Assert.That(task3.task.Result, Is.EqualTo(2));
    }

    [Test]
    public async Task TestChainedTaskDoesNotCompleteBeforeChildTasks()
    {
        var mutex = new ManualResetEventSlim(false);

        var task = taskChain.Add(async () =>
        {
            await Task.Run(() => mutex.Wait(globalCancellationToken.Token)).CatchUnobservedExceptions();
        });

        // Allow task to potentially complete
        Thread.Sleep(1000);

        Assert.That(task.IsCompleted, Is.False);

        // Allow the task to complete.
        mutex.Set();

        await task;
    }

    private (Task<int> task, ManualResetEventSlim mutex, CancellationTokenSource cancellation) addTask()
    {
        var mutex = new ManualResetEventSlim(false);
        var completionSource = new TaskCompletionSource<int>();

        var cancellationSource = new CancellationTokenSource();
        var token = CancellationTokenSource.CreateLinkedTokenSource(cancellationSource.Token, globalCancellationToken.Token);

        taskChain.Add(() =>
        {
            mutex.Wait(globalCancellationToken.Token);
            completionSource.SetResult(Interlocked.Increment(ref currentTask));
        }, token.Token);

        return (completionSource.Task, mutex, cancellationSource);
    }
}
```

peppy committed:
Proposal to use extension method instead of TaskChain class

peppy committed:
Comment and add xmldoc

peppy committed:
Update TaskChain to use ContinueWithSequential internally

It turns out we may still want to use TaskChain for its locking
behaviour, so I've made it internally use the refactored version I
implemented, while keeping the general structure.

peppy committed:
Revert test changes to test original class/scope

Importantly, this removes the call to CatchUnobservedExceptions(), which was
outright incorrect (awaiting on the wrong task as a result) in the
original test code.

peppy commented:
@smoogipoo i've fixed things up with your existing tests and would appreciate a re-review of just the `TaskChain` logic specifically. I'm going to be going over the actual usage of `TaskChain` in the game code next.

Because the extension method I made doesn't handle the locking on chaining of tasks, I've left your `TaskChain` class and made it use `ContinueWithSequence` internally, which tidies things up in its implementation.

peppy reviewed (dismissed):
I think this PR is in a good state to get in. I'm still going to be doing a more thorough review of the surrounding code, but what we have here serves only as an improvement to readability when doing that review.

smoogipoo commented:
The `TaskChain` implementation + extension methods looks good to me.

bdach committed:
Merge branch 'master' into fix-leave-room-race-2

bdach commented:
I've resolved conflicts here, but I can still break the "create room" scenario pretty reliably.

Repro steps:

1. Try to create a room
2. Press <kbd>Esc</kbd> twice after clicking the button that actually creates the room on the settings overlay
3. Re-enter multi
4. Room will still be present
5. Trying to join said room will result in seemingly infinite loading

bdach committed:
Set canceled result in scheduleAsync

Was holding up the task completion source, and in consequence,
potentially the entire task chain.

bdach commented:
Added a fix for the failure scenario I was seeing. Hopefully that addresses the most recent CI fail (since that was a timeout) too...

Still would like a clarification on the above review comment before proceeding, though.

bdach committed:
Don't part room if join task was cancelled

bdach commented:
Fired off another commit to hopefully fix the actual test failures. The automatic part-room-if-join-failed was getting ran on cancelled tasks, which we presumably haven't noticed before because join tasks haven't really been cancelled all that often before. Seems more correct to not clean up in that case.

peppy commented on osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs:
```diff
@@ -142,23 +161,33 @@ public async Task JoinRoom(Room room)
             {
-                if (Room == null)
-                    return;
+                // The join may have not completed yet, so certain tasks that either update the room or reference the room should be cancelled.
+                // This includes the setting of Room itself along with the initial update of the room settings on join.
+                joinCancellationSource?.Cancel();
```
This sounds correct.. @smoogipoo can you check this and confirm intention?

smoogipoo replied:
The room updates from `JoinRoom` are not necessarily scheduled already.

Suppose a `LeaveRoom` happens that schedules before the schedule on [L144](https://github.com/smoogipoo/osu/blob/90ba8ae234bbec60aab5b42f9cd625eb8e2d2f02/osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs#L140-L144) of `JoinRoom`. In this case the room will be set to null by the `LeaveRoom` schedule but then set to non-null by a previous invocation of `JoinRoom`.

As for whether this could be unscheduled... I think so.

smoogipoo commented:
Test::
https://github.com/smoogipoo/osu/blob/90ba8ae234bbec60aab5b42f9cd625eb8e2d2f02/osu.Game/Online/Multiplayer/StatefulMultiplayerClient.cs#L140-L144

smoogipoo committed:
Unschedule cancellation

bdach commented:
`TestSceneParticipantsList` tests failed thrice in a row now, so seems like there's still something wrong. That said, I've tried to debug, and while I could repro at first, now I cannot anymore...

smoogipoo committed:
Fix test failures

smoogipoo commented:
Repro'd by adding a Task.Delay(1000) in `JoinRoom()`.

smoogipoo committed:
Remove unused using
